QUARKUS-1676 Add tests for 'MAIF - mTLS with OIDC Server'

michalvavrik opened the pull request:
This commit add tests for a mutual TLS with OIDC as suggested by [a test plan](https://github.com/quarkus-qe/quarkus-test-plans/pull/76). There are tests for JKS/PKCS12 keystore types, a fail test when a keystore file is completely missing, test with Keycloak and RH SSO 7.3/7.5 in role of a OIDC server and test with keystore file whose file type can't be detected by a file extension.

pjgg commented on security/oidc-client-mutual-tls/src/test/java/io/quarkus/ts/security/oidcclient/mtls/AuthenticationTestSuite.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    /**
+     * Ensures that {@link AuthenticationTestSuite#verifyAuthenticated()} result make sense.
```
This commend should be reviewed

michalvavrik replied:
I removed the comment and added a new test. Now I think context (method name in context of its class) shall clarify method itself.

pjgg commented on security/oidc-client-mutual-tls/src/test/java/io/quarkus/ts/security/oidcclient/mtls/AuthenticationTestSuite.java:
```diff
@@ -0,0 +1,39 @@
+import org.apache.http.HttpStatus;
+import org.junit.jupiter.api.Test;
+
+public interface AuthenticationTestSuite {
```
Could be renamed this interface to something more "meaningful", something like `KeycloakMtlsAuthN`, because all of these names are going to be used by "Polarion" in order to generate a report. 

So if for example `verifyAuthenticated` fails, we are going to have an alert like: 
`KeycloakMtlsAuthN.verifyAuthenticated` fails.

michalvavrik replied:
Refactored, thanks!

pjgg commented on security/oidc-client-mutual-tls/src/test/java/io/quarkus/ts/security/oidcclient/mtls/MutualTlsKeycloakService.java:
```diff
@@ -0,0 +1,71 @@
+import io.quarkus.test.bootstrap.KeycloakService;
+import io.quarkus.test.bootstrap.Protocol;
+
+public class MutualTlsKeycloakService extends KeycloakService {
```
This implementation should be merged into QuarkusTestFramework [keycloakService](https://github.com/quarkus-qe/quarkus-test-framework/blob/main/quarkus-test-service-keycloak/src/main/java/io/quarkus/test/bootstrap/KeycloakService.java). Could you create an issue in [QuarkusTestFramework](https://github.com/quarkus-qe/quarkus-test-framework) in order to add this issue to the backlog.

michalvavrik replied:
Shall do, let's merge this MR first and update it once I've moved `MutualTlsKeycloakService` to QuarkusTestFramework.

pjgg commented on security/oidc-client-mutual-tls/src/test/java/io/quarkus/ts/security/oidcclient/mtls/AuthenticationTestSuite.java:
```diff
@@ -0,0 +1,39 @@
+import org.apache.http.HttpStatus;
+import org.junit.jupiter.api.Test;
+
+public interface AuthenticationTestSuite {
```
I am not sure if this interface should be a class as `IncorrectKsFileTypeOidcMtlsIT`
Maybe could be a class and extends `BaseOidcMtlsIT`

michalvavrik replied:
Refactored, thanks!

pjgg commented on security/oidc-client-mutual-tls/src/test/java/io/quarkus/ts/security/oidcclient/mtls/MutualTlsKeycloakService.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    /**
+     * OpenShift 'passthrough' route listens on 443 and redirects to internal port 8443.
```
typo: passthrough -> passThrough

michalvavrik replied:
Fixed

pjgg commented on security/oidc-client-mutual-tls/src/test/resources/openshift-rh-sso-deployment-template.yml:
```diff
@@ -0,0 +1,77 @@
+---
+apiVersion: "v1"
```
I think that would be great to add this template to [Quarkus-test-framework](https://github.com/quarkus-qe/quarkus-test-framework) so based on `KeycloakService` configuration the framework is able to deploy keycloak with Mtls or a regular one. Please add this feature as part of the previously created issue

michalvavrik replied:
In order to do that, I'll need to adjust at the very least `io.quarkus.test.services.containers.OpenShiftContainerManagedResource#replaceDeploymentContent` as `tls.crt` and `tls.key` should be configurable if we add ths template to Quarkus-test-framework. I will do this change together with the other issue.

pjgg reviewed (changes requested):
Missing add a reference in project `README.md`, that explains what this module does. You could use as an example some of the previous modules ex: `security/basic`

Some of the `test/resources` certificates already exist in `main/resource` folder, so you don't need to duplicate it, just use `main/resource` certificates.

michalvavrik commented:
Hello @pjgg ,

thank you for the review. I added a reference to `README.md`, removed duplicate certificates and resolved issues mentioned in your comments. Regarding suggested QuarkusTestFramework enhancement, let's do it separately once this MR is merged and refactor the test suite afterwards.

BR,
Michal
